File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculations/BodyForce.java
package com.github.aadvorak.artilleryonline.battle.calculations;

import com.github.aadvorak.artilleryonline.battle.common.Force;
import com.github.aadvorak.artilleryonline.battle.common.Position;
import com.github.aadvorak.artilleryonline.battle.common.Vector;
import lombok.AllArgsConstructor;
import lombok.Setter;

@Setter
@AllArgsConstructor
public class BodyForce {

    private Force moving;

    private Force rotating;

    private Vector radiusVector;

    private String description;

    public Force moving() {
        return moving;
    }

    public Force rotating() {
        return rotating;
    }

    public Vector radiusVector() {
        return radiusVector;
    }

    public String description() {
        return description;
    }

    public double torque() {
        if (radiusVector == null || rotating == null) {
            return 0;
        }
        return radiusVector.vectorProduct(rotating);
    }

    public static BodyForce of(Force force, Position position, Position comPosition, String description) {
        var radiusVector = comPosition.vectorTo(position);
        var moving = Force.of(force.projectionOnto(radiusVector));
        var rotating = new Force()
                .setX(force.getX() - moving.getX())
                .setY(force.getY() - moving.getY());
        return new BodyForce(moving, rotating, radiusVector, description);
    }

    public static BodyForce atCOM(Force force, String description) {
        return new BodyForce(force, null,null, description);
    }

    @Override
    public String toString() {
        return "BodyForce{" +
                "moving=" + moving +
                ", rotating=" + rotating +
                ", radiusVector=" + radiusVector +
                ", description='" + description + '\'' +
                '}';
    }
}

File: frontend/playground/battle/calculator/body-force.ts
import {type Force, type Position, type Vector, zeroVector} from "~/playground/data/common";
import {VectorUtils} from "~/playground/utils/vector-utils";

export class BodyForce {
  moving: Force | null
  rotating: Force | null
  radiusVector: Vector | null
  description: string

  constructor(
      moving: Force | null,
      rotating: Force | null,
      radiusVector: Vector | null,
      description: string
  ) {
    this.moving = moving
    this.rotating = rotating
    this.radiusVector = radiusVector
    this.description = description
  }

  torque(): number {
    if (!this.radiusVector || !this.rotating) {
      return 0
    }
    return VectorUtils.vectorProduct(this.radiusVector, this.rotating)
  }

  sumForce(): Force {
    return VectorUtils.sumOf(this.moving || zeroVector(), this.rotating || zeroVector())
  }

  sumForceMagnitude(): number {
    let magnitude = 0
    if (this.rotating) {
      magnitude += VectorUtils.getMagnitude(this.rotating)
    }
    if (this.moving) {
      magnitude += VectorUtils.getMagnitude(this.moving)
    }
    return magnitude
  }

  static of(
      force: Force,
      position: Position,
      comPosition: Position,
      description: string
  ): BodyForce {
    const radiusVector = VectorUtils.vectorFromTo(comPosition, position)
    const moving = VectorUtils.projectionOfOnto(force, radiusVector)
    const rotating = {
      x: force.x - moving.x,
      y: force.y - moving.y
    }
    return new BodyForce(moving, rotating, radiusVector, description)
  }

  static atCOM(force: Force, description: string): BodyForce {
    return new BodyForce(force, null, null, description)
  }
}